### PR TITLE
SIGUSR1 to reload certificates

### PR DESCRIPTION
**Pull request by RemyNoulin**
There is something missing in my patch.

When I reload the certificates, I get an error in the browser:
Error code: SEC_ERROR_BAD_SIGNATURE

I run with these options: kore -fnrdc, so chroot is not activated.

I don't know if the problem is in the keymgr or the worker.

I will continue debugging this, something related to openssl has to be reinitialized.

### CODE REVIEW

**Comment by jorisvink** [comment]
You aren’t clearing the list of private keys before loading the new ones.

Which means if you swap keys and reload the old key is still going to be used.

**Comment by jorisvink** [comment]
Also you'll want to be able to give the keymgr process a chroot path of its own.

File: src/keymgr.c [hunk 1]
```diff
@@ -106,6 +106,13 @@ kore_keymgr_run(void)
 			case SIGTERM:
 				quit = 1;
 				break;
+			case SIGUSR1: {
```
Comment (jorisvink) [inline src/keymgr.c]: Gross. Please don't do this. Use the same style as everywhere else.

**Comment by jorisvink** [comment]
Ideally I want something like this:

```
runas nobody
worker_cert_path /path/to/certfiles

keymgr_runas _keymgr
keymgr_key_path /path/to/keyfiles

domain kore.io {
    certfile relative/path/inside/worker_cert_path/cert.pem
    keyfile relative/path/inside/keymgr_key_path/key.pem
}
```

So there is a clear separation between file access of key files and certificate files for keymgr and worker processes respectively.

File: src/domain.c [hunk 2]
```diff
@@ -760,4 +769,15 @@ domain_x509_verify(int ok, X509_STORE_CTX *ctx)
 
 	return (ok);
 }
+
+void
+kore_domain_reload_certificates(void)
+{
+	static struct kore_domain	*dom;
```
Comment (jorisvink) [inline src/domain.c]: why on earth is this stack variable declared static?

File: src/worker.c [hunk 3]
```diff
@@ -219,22 +219,24 @@ kore_worker_dispatch_signal(int sig)
 }
 
 void
-kore_worker_privdrop(void)
+kore_worker_privdrop(int id)
```
Comment (jorisvink) [inline src/worker.c]: no need, worker 0 when TLS is enabled will always be keymgr process. So you can check worker->id.

File: src/worker.c [hunk 4]
```diff
@@ -376,6 +378,12 @@ kore_worker_entry(struct kore_worker *kw)
 			case SIGTERM:
 				quit = 1;
 				break;
+			case SIGUSR1:
+				kore_domain_reload_certificates();
+				kore_domain_load_crl();
+				kore_domain_keymgr_init();
```
Comment (jorisvink) [inline src/worker.c]: This shouldn't have to be done for each reload.

File: src/worker.c [hunk 5]
```diff
@@ -219,22 +219,24 @@ kore_worker_dispatch_signal(int sig)
 }
 
 void
-kore_worker_privdrop(void)
+kore_worker_privdrop(int id)
 {
 	rlim_t			fd;
 	struct rlimit		rl;
 	struct passwd		*pw = NULL;
+	char			*user;
```
Comment (jorisvink) [inline src/worker.c]: This can be a const pointer.

File: src/worker.c (hunk 3 above)
Comment (RemyNoulin) [reply src/worker.c]: done

**Comment by RemyNoulin** [comment]
The key and certificate are loaded in the chrooted environment, before they were loaded while parsing the config file. When chroot is active, certfile and certkey are relative to / in the chrooted environment.

**Comment by jorisvink** [comment]
There are still outstanding issues that should be resolved before I am looking to merge this.

File: src/domain.c (hunk 2 above)
Comment (RemyNoulin) [reply src/domain.c]: done. (I'm trying to find which issue is outstanding, then I will rebase the patch on latest)

File: src/worker.c (hunk 5 above)
Comment (RemyNoulin) [reply src/worker.c]: done

File: src/worker.c [hunk 6]
```diff
@@ -219,22 +219,24 @@ kore_worker_dispatch_signal(int sig)
...
 
 	/* Must happen before chroot. */
 	if (skip_runas == 0) {
-		pw = getpwnam(runas_user);
+		user = id == KORE_WORKER_KEYMGR ? runkeymgras_user : runas_user;
```
[the remark this answers is not in the document]
Comment (RemyNoulin) [reply src/worker.c]: Handled in config.c

**Comment by jorisvink** [comment]
A very different version of this was commited in https://github.com/jorisvink/kore/commit/cffb7ec3798684c3a59326de9c1e1f286db34334 - closing this PR.

**Pull request closed without merging**